AuthContext, axiosinstance 추가

seongjin0320 opened the pull request:
# AuthContext, axiosinterceptor 추가

## 설명

- **이 PR은 어떤 종류의 수정 사항인가요?** (버그 수정, 기능, 문서 업데이트 등)
- api 인증 상태 관리 기능 추가

- **상세 설명**
- 로그인된 상태를 전역적으로 관리하기 위해 AuthContext.tsx 파일을 통해 전역상태로 인증상태를 관리할 수 있도록 했습니다.
- axiosinstance.ts를 통해 모든 HTTP요청을 중앙화 해 관리할 수 있게 했습니다. (에러코드에 따른 에러핸들링 로직 추가 해야할 수 있음)
- AuthContext.tsx에서 사용할 유저 타입을 src/pages/space/types/member.ts 를 재사용하기 위해 src/types/member.ts로 파일 위치를 옮겼습니다. 이에따른 Profile.tsx 파일의 import문 수정이 있었습니다.
- AuthProvider를 사용하기 위해 main.tsx에 코드 추가가 있었습니다.

- **기타 정보**:
- 아직 로그인/회원가입 API요청(POST) 부분은 구현하지 않았습니다.
- app.tsx를 수정해 AuthContext를 통해 localStorage에 토큰을 저장하여 로그인 상태를 유지시키는것과, login(), logout()이 작동하는것을 확인했습니다. 
- 이상한 부분이 있으면 적극적으로 알려주시면 좋아요.. 저의 이해도가 낮은 부분이라..

<!--- 바로 아래에 이슈 링크를 추가해 주세요.-->

seongjin0320 committed:
chore: axios, tanstack/react-query 라이브러리 추가

seongjin0320 committed:
feat: 인증정보 전역컨텍스트, 인증 토큰 interceptor 생성

- 인증상태를 다루기 위한 전역 컨텍스트를 src/contexts/AuthContext.tsx 파일로 구현했습니다
- src/api/axiosInstance.ts 파일을 통해 API 클라이언트를 중앙화 했습니다
- AuthProvider와 QueryClientProvider를 main.tsx에 추가해 인증관련 API처리 기능을 구현했습니다
- 아직 로그인, 회원가입 API요청 부분은 추가하지 않았습니다

seongjin0320 committed:
feat: AuthContext.tsx 의 user 타입 member.ts 재사용

- 개인스페이스에서 사용하던 memberType을 재사용하기위해 member.ts를 src/types 공통 타입 폴더 내부로 이동시켰습니다

Moderator11 commented on src/contexts/AuthContext.tsx:
```diff
@@ -0,0 +1,73 @@
+
+// AuthContext에서 사용할 User 타입. MemberType을 기준으로 삼습니다.
+
+type User = MemberType;
```
오 MemberType 기억해주시고 재사용 하셨네요! 정말 좋습니닷 :)

Moderator11 commented on src/contexts/AuthContext.tsx:
```diff
@@ -0,0 +1,73 @@
+  logout: () => void;
+}
+
+const AuthContext = createContext<AuthContextType | undefined>(undefined);
```
AuthContext를 null 대신 undefined로 초기화 하는 이유는 무엇인가요? (순수 궁금증)

seongjin0320 replied:
사실 이렇게 추천하길래 쓴거긴한디 찾아보니까
AuthContext의 기본값이 없기 때문에, 기본값을 undefined로 두고, useAuth에서 undefined가 뜨면 에러를 던지기 위해서, 
그리고 null은 로그아웃 상태임을 나타내는 등으로 많이 쓰인다고 하여 피했고, 타입 안정성 측면에서도 undefined가 더 좋다고 하네요...

Moderator11 commented on src/contexts/AuthContext.tsx:
```diff
@@ -0,0 +1,73 @@
+  useEffect(() => {
+    // 앱이 처음 로드될 때 localStorage에서 토큰과 사용자 정보를 가져와 상태를 초기화합니다
+    const storedToken = localStorage.getItem('authToken');
+    const storedUser = localStorage.getItem('authUser');
```
localStorage에 사용되는 key 값도 상수화 해주면 좋을 것 같아요 :)
localStorage는 서비스 전체에서 활용되는 공간이라 혹시나 키 값이 겹칠 일을 방지 할 수 있을 것 같아서요!

seongjin0320 replied:
src/utils/storageKeys.ts 파일로 분리하여 상수화 했습니다!

Moderator11 commented on src/contexts/AuthContext.tsx:
```diff
@@ -0,0 +1,73 @@
+    throw new Error('useAuth must be used within an AuthProvider');
+  }
+  return context;
+};
```
로직을 보면 context가 없을 때 (인증이 이루어지지 않았을 때?) Error을 반환하는데, Error 메세지는 훅이 AuthProvider 안에서 사용되어야 한다는 내용이에요.

"로그인이 필요합니다"와 비슷한 맥락의 메세지는 어떨까요?

그리고 useAuth라는게 훅인 것 같은데 hook 폴더 말고 context 안에 들어있네용

제가 잘못 이해했다면 말씀주셔용

seongjin0320 replied:
context가 없는 상황은 인증이 이루어지지 않은 상황보다는 useAuth가 AuthProvider 내부에서 사용되지 않았을 때 생기는 상황인것 같아요! 그래서 그런 상황이 발생했을 때 개발자가 문제를 찾을 수 있도록 삽입된 에러 메시지이므로, "로그인이 필요합니다" 라는 메시지같은 유저를 위한 메시지는 실제 로그인 관련 로직으로 핸들링 하면 좋을것 같다고 생각했어용 

useAuth는 src/hooks/useAuth.ts 파일로 분리했습니다!

Moderator11 commented on src/api/axiosInstance.ts:
```diff
@@ -0,0 +1,40 @@
+    }
+    return Promise.reject(error);
+  }
+);
```
👍 👍

Moderator11 reviewed (changes requested):
전반적으로 잘 구현해 주셨네요! 그런데 지금 토큰을 담는 변수가 `token` 하나밖에 없어요.
FE 멘토링 첫 번째 시간에도 나왔었던 내용인데, JWT를 사용한 로그인 방식에서는 대부분 accessToken이외에도 refreshToken이 함께 날라와요.

- accessToken은 백엔드에서 인가 기능에 사용하기 위해 FE쪽에서 헤더에 실어 보내는 토큰이에용
- refreshToken은 accessToken이 만료되었을 때 accessToken을 재발급 받기 위한 토큰입니당

그래서 토큰을 저장할 자리가 총 2개가 필요해요! 그 부분 수정 부탁드립니다~
나머지 피드백은 선택적으로 반영해주세용

수고하셨습니다!

seongjin0320 committed:
refact: 인증 컨텍스트에서 authToekn사용하던것 accessToken, refreshToken을 사용하는 방식으로 변경

- axiosInstance.ts에서 모든 요청에 대해 accessToken을 자동으로 헤더로 붙이도록 했습니다
- 401에러(만료/ 인증실패) 시에 refresh시도 후 요청을 재시도하도록 했습니다
- 리프레시 엔드포인트, 리프레시토큰 반환 형태를 정확히 몰라 그 부분은 백엔드에서 제공하는 형태로 변경 필요할 수 있습니다
- localStorage에 접근하기 위한 localStorage키는 src/utils/storageKeys.ts로 분리하여 상수화 했습니다
- axiosInstance의 토큰 변경을 감지하기 위해 브라우저 이벤트 리스너를 설정해 context의 상태를 동기화 할 수 있도록 했습니다
- AuthContext.tsx 내부에 useAuth 훅을 src/hooks/ 내부로 분리했습니다

seongjin0320 commented:
- axiosInstance.ts에서 모든 요청에 대해 accessToken을 자동으로 헤더로 붙이도록 했습니다
- 401에러(만료/ 인증실패) 시에 refresh시도 후 요청을 재시도하도록 했습니다
- 리프레시 엔드포인트, 리프레시토큰 반환 형태를 정확히 몰라 그 부분은 백엔드에서 제공하는 형태로 변경 필요할 수 있습니다
- localStorage에 접근하기 위한 localStorage키는 src/utils/storageKeys.ts로 분리하여 상수화 했습니다
- axiosInstance의 토큰 변경을 감지하기 위해 브라우저 이벤트 리스너를 설정해 context의 상태를 동기화 할 수 있도록 했습니다
- AuthContext.tsx 내부에 useAuth 훅을 src/hooks/ 내부로 분리했습니다

피드백 반영 내용입니다!

Moderator11 commented on src/utils/storageKeys.ts:
오... . 으로 네임스페이스를 구분하셨네요, 이렇게 해볼 생각은 한번도 못했는데, storage 값들이 많아졌을 때, 디버깅이 편할 것 같네욥 👍

Moderator11 commented on src/contexts/AuthContext.tsx:
```diff
@@ -1,73 +1,91 @@
       }
     }
+
+    // 동일 탭 내에서 axiosInstance의 토큰 변경을 감지하기 위한 브라우저 이벤트 리스너 설정
+    const handleRefreshed = (e: Event) => {
+      const detail = (
+        e as CustomEvent<{ accessToken: string | null; refreshToken?: string | null }>
+      ).detail;
+      if (detail?.accessToken !== undefined) setAccessToken(detail.accessToken);
+      if (detail?.refreshToken !== undefined && detail.refreshToken !== null)
+        setRefreshToken(detail.refreshToken);
+    };
+    const handleCleared = () => {
+      setAccessToken(null);
+      setRefreshToken(null);
+      setUser(null);
+    };
+    window.addEventListener('auth:tokenRefreshed', handleRefreshed as EventListener);
+    window.addEventListener('auth:tokensCleared', handleCleared);
+    return () => {
+      window.removeEventListener('auth:tokenRefreshed', handleRefreshed as EventListener);
+      window.removeEventListener('auth:tokensCleared', handleCleared);
+    };
```
커스텀 이벤트를 구현해본 적이 없어서 이부분은 이해가 안가서 설명해주시면 감사하겠습니닷

auth:tokenRefreshed, auth:tokensCleared 로직을 이벤트로 작성하신 이유가 무엇인가요?
특히 clear 동작에서 지금 보면 axiosInstance에서 storage 값은 함수 호출로 삭제하지만, 내부 state 값은 이벤트 핸들러 Dispatch로 삭제하는데, 이렇게 작성하신 이유가 궁금해욥

Moderator11 reviewed (approved):
안녕하세요 성진님!
피드백 모두 잘 수용해주셨고, 일부분에서는 로직 개선도 보인 것 같아요!
일부 코드를 이해하지 못해서 질문 남겼는데, 답변 부탁드립니다.
수고하셨습니다. 👍
